### ISSUE TITLE
Gradient Descent: stopping condition and other problems

### ISSUE THREAD

**Issue by WiIIiamTang** [WiIIiamTang/term_project_360-420-w2019-section2-Tang-Dinh#11]
**LogisticWine1**

Currently, the gradient descent loop will break if 1) the max number of iterations are reached, or 2) the "new" cost function evaluated at the updated parameters is larger than the previous cost function.

The beta array is randomly initialized each time the program is run by assigning to each beta a value between 0 and 1.

Issues: 

1. In some rare instances, the loop is immediately terminated, or stopped after only a few thousand iterations, while in other trials, the loop reaches it's max number of iterations without problem. *refer to the "does not converge" issue given a temporary fix earlier*

2. In a follow-up to 1, the stopping condition for our gradient descent may not be optimal

3. Because of this and probably some other things, the accuracy gets really weird; the results are anywhere from 70 - 90%

4. Lowering the alpha, even to very small values, does *not* fix this issue (1, 2, 3); increasing the alpha to get faster times results in an increase of issue(1) and decreasing alpha seems to have no effect in improving the accuracy nor in solving issue(1)

5. Testing the same implementation of logistic regression on the breast cancer data showed really good results. So why isn't it working here?

**Comment by WiIIiamTang** [comment]
Possible cause found:

In testing the logistic regression against another data set, this time much simpler (2 dimensions, 100 points), the results were really good. We used the data from some school exam marks and the admittance to university. In a ML workshop about a month ago, we used this data to implement logistic regression in python. When we used the java model we made with the same data, we got practically _the same numbers._ This indicates to us that our gradient descent algorithm is not the problem.

However, in this test, we did not shuffle our data around (the shuffleData method was not used). Also, if we changed the **test/training** data split to something other than 70%, the whole thing would blow up. 

Next, with these observations in mind, we tested the breast cancer data again, but this time without the shuffleData method. Results are still really good (90+), and more importantly the cost function does not increase near the start or after a few iterations. Also, if we changed the test/training data split (eg. to 70%), this would again blow up the whole program.

Another thing to note is that with both these test cases, the gradient descent would not work without normalization.

I personally think that this could mean there are maybe three things causing all of the issues raised in the original post:

1. the shuffleData method does not work as intended, and/or

2. depending on what data points are in the training/test sets, the initial cost function and gradient are accordingly influenced; in some cases, it just so happens that the data points in the training set give values that are not appropriate for gradient descent, and/or

3. the normalization of data is erroneous.
